### PR TITLE
[Experimental] Product filters: Add insertion feature

### PR DESCRIPTION
**Pull request by roykho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds the ability for filters to be added back within the parent wrapper container.

NOTE:

- I also took this opportunity to update the API version to 3 to align with other filters.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Test by developers only**

Prereq:
- Ensure you have the WC Beta Tester Plugin installed and go to Tools > WCA Test Helper > Features. Toggle on the experimental-blocks option.

1. Go to the editor and navigate to Product Catalog template.
2. Add the Product Filters (Experimental) block to the editor.
3. Delete any of the filter (Active,Status,Price,Attributes,Rating).
4. Within the Product Filters Wrapper, use the inserter + button and try to add back the filters.
5. Ensure you can add each one of the filters back.
6. Try adding more than one of the same filter. You should see a warning stating one has already be added. Attributes filter is the exception. We will allow multiples of this filter.
7. Ensure you still can't insert any of the individual filters outside of the wrapper block.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 
Add ability to add individual filters back
</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts [hunk 1]
```diff
@@ -507,4 +507,38 @@ test.describe( `${ blockData.name }`, () => {
...
+			'price (experimental)',
+			'rating (experimental)',
+			'attribute (experimental)',
+			'active (experimental)',
+		];
+
+		const addBlockButton = block.getByRole( 'button', {
+			name: 'Add block',
+		} );
+		await addBlockButton.dispatchEvent( 'click' );
```
Comment (thealexandrelara) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts]: Here you can use the `click` method instead:

```suggestion
		await addBlockButton.click();
```

**Review by thealexandrelara — CHANGES_REQUESTED** [review]
Thank you for working on this. I left a small comment about the code.

✅ The individual filters: Price, Rating, Stock Status, Attribute can be aded to the Product Filters block;
✅ When the Price, Rating, Stock Status are added more than once, a message is displayed to show users that these blocks can be added only once and giving them some options.

---
❌  When a block is added more than once, and the message is displayed, the Remove button is not working.

https://github.com/user-attachments/assets/1b023bb9-a6fc-4ffb-8acf-5aeeea58dfc2

---
❌  When the Attribute filter block is added more than once, a message is displayed informing users that it can only be added once. The correct behavior should allow merchants to add multiple Attribute filter blocks. For instance, they should be able to enable shoppers to filter by both Size and Color.

**Comment by roykho** [comment]
@thealexandrelara thanks for the review please see my note:

> When a block is added more than once, and the message is displayed, the Remove button is not working.

That is a GB bug.

> When the Attribute filter block is added more than once, a message is displayed informing users that it can only be added once. The correct behavior should allow merchants to add multiple Attribute filter blocks. For instance, they should be able to enable shoppers to filter by both Size and Color.

Right. I forgot they would have to add multiple for different attributes.

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
Thank you for working on this PR. Here is my testing result following the instructions:

> Ensure you can add each one of the filters back.
> Ensure you still can't insert any of the individual filters outside of the wrapper block.

I can confirm these items.
 
> > > Try adding more than one of the same filter. You should see a warning stating one has already be added. Attributes filter is the exception. We will allow multiples of this filter.

> > When a block is added more than once, and the message is displayed, the Remove button is not working.

> That is a GB bug.

I can confirm the Remove button is not working as @thealexandrelara mentioned above. But I don't think it's totally a GB bug. If you unlock the Price Filter block in Alexandre's example, you can remove it (only the Filter Options). We're adding the block variations of `product-filter`, not the filter options blocks, so we may need a custom handler here.

**Comment by roykho** [comment]
> I can confirm the Remove button is not working as @thealexandrelara mentioned above. But I don't think it's totally a GB bug. If you unlock the Price Filter block in Alexandre's example, you can remove it (only the Filter Options). We're adding the block variations of product-filter, not the filter options blocks, so we may need a custom handler here.

But the lock feature and remove feature are from GB and we didn't add anything custom here. I think GB should gracefully handle this and say "Please unlock your block before clicking remove". wdyt?

**Comment by thealexandrelara** [comment]
> But the lock feature and remove feature are from GB, and we didn't add anything custom here. I think GB should gracefully handle this and say "Please unlock your block before clicking remove". wdyt?

I think that what Tung is saying is that when we click the Remove button, it tries to remove the inner filter block (the "Filter Options" block within the Product Filter block - which is locked), so we should implement a custom handler to remove the Product Filter block instead (the Price Filter). This one:

CleanShot 2024-07-22 at 16 52 23@2x

---

Another thing I noticed is that I can add multiple Active Filters blocks to the Product Filters block:

https://github.com/user-attachments/assets/c53d4886-d029-4eea-b88e-859fa5949493

**Comment by roykho** [comment]
@thealexandrelara do you have an idea on how to override the core removal behavior?

**Comment by dinhtungdu** [comment]
One option is dropping the `product-filter` and its variations and creating wrapper blocks for each filter. By doing so, we can control the number of instance through `multiple` support. It's also easier to share the inspector settings between wrapper and filter option blocks.

**Comment by thealexandrelara** [comment]
Yeah, I believe that Tung's suggestion would be the best way to do that because, currently, the variations are making it harder to modify each block. For instance, one thing I noticed is that when we change the block's name, it only changes in the list view, not in the InspectorControls sidebar, and I don't know if this is possible using block variations, so we may need to explore that. I'll have to create an issue to start investigating this

**Comment by roykho** [comment]
@dinhtungdu @thealexandrelara - so basically the suggestion is to undo all the work that was done [here](https://github.com/woocommerce/woocommerce/pull/48638) and don't use variations but just have unique wrappers for each individual filter?

**Comment by dinhtungdu** [comment]
> so basically the suggestion is to undo all the work that was done [here](https://github.com/woocommerce/woocommerce/pull/48638) and don't use variations but just have unique wrappers for each individual filter?

No, it's not all the work, just the part related to the `product-filter` block is removed.

Thinking more about this, I don't think it's worth trading product variation to block insertion limitation. I think we can have both (using `product-filter` variation to avoid duplicated wrapper blocks and meet the block insertion specs). There are two problems need to solve:
1. Build a custom checker, similar (but simplified) to block multiple supports, in the `product-filter` block, check if there is any block inside `product-filters` block that doesn't support multiple instance, then add a warning with a button to remove the newly inserted block.
2. Investigate to share the block controls between filter option blocks and the `product-filter` block.

@thealexandrelara I see you created #49811, I think we should at least tried achieving two points above with current block structure, as creating each wrapper block for each filter block resulting a lot of duplication.

**Comment by thealexandrelara** [comment]
> @thealexandrelara I see you created https://github.com/woocommerce/woocommerce/issues/49811, I think we should at least tried achieving two points above with current block structure, as creating each wrapper block for each filter block resulting a lot of duplication.

Yeah, I created that issue just to keep track of things, but it is not prioritized yet. I added it to be discussed for future cycles. If we manage to solve both problems that you raised, that issue won’t be needed anymore, and we can close it. I agree with you that we should first try solving those while keeping the current block structure.

**Comment by roykho** [comment]
Closing this PR as it's been stale too long. Will create a new PR for this.

**Pull request closed without merging**